test/ILLink.RoslynAnalyzer.Tests/TestCaseUtils.cs

```diff
@@ -34,106 +35,240 @@ public static IEnumerable<object[]> GetTestData (string testSuiteName)
-								var secondColon = text.IndexOf (": ", firstColon + 1);
-								if (secondColon > 0) {
-									text = text.Substring (secondColon + 2);
-								}
-							}
-						}
-						bool found = false;
-						foreach (var d in filtered) {
-							if (d.Contains (text)) {
-								found = true;
-								break;
-							}
-						}
-						if (!found) {
-							var diagStrings = string.Join (Environment.NewLine, filtered);
-							Assert.True (false, $@"Could not find text:
+				}
+			}
+
+			if (!found) {
+				var diagStrings = string.Join (Environment.NewLine, diagnostics);
+				Assert.True (false, $@"Could not find text:
 {text}
 In diagnostics:
 {diagStrings}");
-						}
-					}
-					break;
-				case "LogDoesNotContain": {
-						var arg = Assert.Single (attr.ArgumentList!.Arguments);
-						var text = GetStringFromExpr (arg.Expression);
-						foreach (var d in filtered) {
-							Assert.DoesNotContain (text, d);
-						}
+			}
+
+			diagnostics.RemoveAt (matchedDiagnosticIndex);
+		}
+
+		private static void ValidateLogDoesNotContainAttribute (AttributeArgumentListSyntax attributeArguments, List<string> diagnostics)
+		{
+			Assert.True (attributeArguments.Parent is AttributeSyntax attr && attr.Name.ToString () == "LogDoesNotContain");
+			var arg = Assert.Single (attributeArguments.Arguments);
+			var text = GetStringFromExpression (arg.Expression);
+			foreach (var diagnostic in diagnostics)
+				Assert.DoesNotContain (text, diagnostic);
+		}
+
+		private static void ValidateUnrecognizedReflectionAccessPatternAttribute (
+			SemanticModel semanticModel,
+			AttributeArgumentListSyntax attributeArguments,
+			ref List<string> diagnostics)
+		{
+			Assert.True (attributeArguments.Parent is AttributeSyntax attr && attr.Name.ToString () == "UnrecognizedReflectionAccessPattern");
+			string expectedWarningCode = string.Empty;
+			if (TryGetNamedArgument (attributeArguments.Arguments!, "messageCode", out var messageCodeArgument)) {
+				var warningCode = GetStringFromExpression (messageCodeArgument!.Expression);
+				Assert.True (warningCode.StartsWith ("IL"),
+					$"The warning code specified in {attributeArguments.ToString ()} must start with the 'IL' prefix. Specified value: '{warningCode}'");
+
+				expectedWarningCode = warningCode;
+			}
+
+			string returnTypeName = string.Empty;
+			if (TryGetNamedArgument (attributeArguments.Arguments!, "returnType", out var returnTypeArgument))
+				returnTypeName = semanticModel.GetTypeInfo ((TypeOfExpressionSyntax) returnTypeArgument!.Expression).Type!.GetDisplayName ();
+
+			MemberDeclarationSyntax sourceMember = attributeArguments.Ancestors ().OfType<MemberDeclarationSyntax> ().First ();
+			if (semanticModel.GetDeclaredSymbol (sourceMember) is not ISymbol memberSymbol)
+				return;
+
+			string expectedSourceMemberName = memberSymbol!.GetDisplayName ();
+			// Build the member signature to match.
+			var sb = new StringBuilder ();
+			if (string.IsNullOrEmpty (returnTypeName))
```

Comment:
It should be the opposite logic, append the returnTypeName when not nullOrEmpty
```suggestion
			if (!string.IsNullOrEmpty (returnTypeName))
```